chirp/drivers/baofeng_uv17.py

```diff
@@ -0,0 +1,423 @@
+               [b"\02", 8], [b"\x06", 1]]
+    _fingerprint = b"\x06" + b"UV15999"
+    _scode_offset = 1
+
+    _tri_band = False
+    POWER_LEVELS = [chirp_common.PowerLevel("Low", watts=1.00),
+                    chirp_common.PowerLevel("High",  watts=5.00)]
+
+    LENGTH_NAME = 11
+    VALID_BANDS = [baofeng_uv17Pro.UV17Pro._vhf_range,
+                   baofeng_uv17Pro.UV17Pro._uhf_range]
+    SCODE_LIST = ["%s" % x for x in range(1, 16)]
+    SQUELCH_LIST = ["Off"] + list("123456789")
+    LIST_POWERON_DISPLAY_TYPE = ["Full", "Message", "Voltage"]
+    LIST_TIMEOUT = ["Off"] + ["%s sec" % x for x in range(15, 615, 15)]
+    LIST_VOICE = ["Chinese", "English"]
+    LIST_BACKLIGHT_TIMER = ["Always On"] + ["%s sec" % x for x in range(1, 11)]
+    LIST_MODE = ["Name", "Frequency"]
+    CHANNELS = 999
+
+    MEM_FORMAT = """
+    #seekto 0x0030;
+    struct {
+      lbcd rxfreq[4];
+      lbcd txfreq[4];
+      u8 unused1;
+      ul16 rxtone;
+      ul16 txtone;
+      u8 unknown1:1,
+         bcl:1,
+         pttid:2,
+         unknown2:1,
+         wide:1,
+         lowpower:1,
+         unknown:1;
+      u8 scode:4,
+         unknown3:3,
+         scan:1;
+      u8 unknown4;
+    } memory[1002];
+
+    #seekto 0x7000;
+    struct {
+      u8 powerondistype;
+      u8 unknown0[15];
+      char boottext1[10];
+      u8 unknown1[6];
+      char boottext2[10];
+      u8 unknown2[22];
+      u8 tot;
+      u8 squelch;
+      u8 vox;
+      u8 powersave: 4,
+         unknown3:2,
+         voice: 1,
+         voicesw: 1;
+      u8 backlight;
+      u8 beep:1,
+         autolock:1,
+         unknown4:1,
+         tail:1,
+         scanmode:2,
+         dtmfst:2;
+      u8 unknown5:1,
+         dualstandby:1,
+         roger:1,
+         unknown6:3,
+         fmenable:1,
+         unknown7:1;
+      u8 unknown8[9];
+      u8 unknown9:6,
+         chbdistype:1,
+         chadistype:1;
+    } settings;
+
+    struct vfo {
+      lbcd rxfreq[4];
+      lbcd txfreq[4];
+      u8 unused1;
+      ul16 rxtone;
+      ul16 txtone;
+      u8 unknown1:1,
+         bcl:1,
+         pttid:2,
+         unknown2:1,
+         wide:1,
+         lowpower:1,
+         unknown:1;
+      u8 scode:4,
+         unknown3:3,
+         scan:1;
+      u8 unknown4;
+    };
+
+    #seekto 0x0010;
+    struct {
+      struct vfo a;
+      struct vfo b;
+    } vfo;
+
+    #seekto 0x4000;
+    struct {
+      char name[11];
+    } names[999];
+
+    #seekto 0x8000;
+    struct {
+      u8 code[5];
+    } pttid[15];
+
+    struct {
+      u8 unknown[5];
+      u8 code[5];
+    } ani;
+
+    """
+
+    def _make_frame(self, cmd, addr, length, data=""):
+        """Pack the info in the header format"""
+        frame = struct.pack("<cH", cmd, addr) + struct.pack(">H", length)
+        # add the data if set
+        if len(data) != 0:
+            frame += data
+        # return the data
+        return frame
+
+    def _download(radio):
```

Comment:
You moved all of this into the radio class, but then override it in the subclass anyway. Are the download routines re-usable or not? If not, there's no reason to move them into the class, and it's generally preferable to keep class sizes down, when possible. Things that need not be overridden can be module-scoped to keep the classes small.

Was there a reason for this or was it just to avoid the reuse heuristic? If the latter, I'm sorry you did that, but it wasn't necessary. That check is trying to prevent copying code (or copying with just a few changes), but we need not jump through hoops that don't otherwise make sense to make it pass.

Reply:
I did struggle a lot with the reuse heuristic. :-) This was to prevent the syn_in method twice. I managed to move it outside of the class now.